crypto/src/hash/rescue_prime/mod.rs

```diff
@@ -0,0 +1,816 @@
+        }
+    }
+
+    fn generate_circulant_matrix(mds_vector: &[Fp]) -> Vec<Vec<Fp>> {
+        let m = mds_vector.len();
+        (0..m)
+            .map(|i| {
+                mds_vector
+                    .iter()
+                    .cycle()
+                    .skip(m - i)
+                    .take(m)
+                    .cloned()
+                    .collect::<Vec<Fp>>()
+            })
+            .collect()
+    }
+
+    pub fn apply_sbox(state: &mut [Fp]) {
+        for x in state.iter_mut() {
+            *x = x.pow(ALPHA);
+        }
+    }
+
+    fn mds_matrix_vector_multiplication(&self, state: &[Fp]) -> Vec<Fp> {
+        let m = state.len();
+        let mut new_state = vec![Fp::zero(); m];
+
+        for (i, new_value) in new_state.iter_mut().enumerate() {
+            for (j, state_value) in state.iter().enumerate() {
+                *new_value += self.mds_matrix[i][j] * state_value;
+            }
+        }
+
+        new_state
+    }
+    fn mds_ntt(&self, state: &[Fp]) -> Vec<Fp> {
+        let m = state.len();
+        let omega = match m {
+            12 => Fp::from(281474976645120u64),
+            16 => Fp::from(17293822564807737345u64),
+            _ => panic!("Unsupported state size for NTT"),
+        };
+
+        let mds_ntt = ntt(&self.mds_vector, omega);
+        let state_rev: Vec<Fp> = iter::once(state[0])
+            .chain(state[1..].iter().rev().cloned())
+            .collect();
+        let state_ntt = ntt(&state_rev, omega);
+
+        let mut product_ntt = vec![Fp::zero(); m];
+        for i in 0..m {
+            product_ntt[i] = mds_ntt[i] * state_ntt[i];
+        }
+
+        let omega_inv = omega.inv().unwrap();
+        let result = intt(&product_ntt, omega_inv);
+
+        iter::once(result[0])
+            .chain(result[1..].iter().rev().cloned())
+            .collect()
+    }
+
+    fn mds_karatsuba(&self, state: &[Fp]) -> Vec<Fp> {
+        let m = state.len();
+        let mds_rev: Vec<Fp> = iter::once(self.mds_vector[0])
+            .chain(self.mds_vector[1..].iter().rev().cloned())
+            .collect();
+
+        let conv = karatsuba(&mds_rev, state);
+
+        let mut result = vec![Fp::zero(); m];
+        result[..m].copy_from_slice(&conv[..m]);
+        for i in m..conv.len() {
+            result[i - m] += conv[i];
+        }
+
+        result
+    }
+
+    fn apply_mds(&self, state: &mut [Fp]) {
+        let new_state = match self.mds_method {
+            MdsMethod::MatrixMultiplication => self.mds_matrix_vector_multiplication(state),
+            MdsMethod::Ntt => self.mds_ntt(state),
+            MdsMethod::Karatsuba => self.mds_karatsuba(state),
+        };
+        state.copy_from_slice(&new_state);
+    }
+
+    fn add_round_constants(&self, state: &mut [Fp], round: usize) {
+        let m = self.m;
+        let round_constants = &self.round_constants;
+        for j in 0..m {
+            state[j] += round_constants[round * 2 * m + j];
```

Comment:
instead of indexing, this can be improved by using get_mut and an iterator

Reply:
updated the implementation as suggested